pkg/apis/pipeline/v1/matrix.go

```diff
@@ -0,0 +1,80 @@
+/*
+Copyright 2023 The Tekton Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+// FanOut produces combinations of Parameters of type String from a slice of Parameters of type Array.
+func (m *Matrix) FanOut() []Params {
+	var combinations []Params
+	for _, parameter := range m.Params {
+		combinations = fanOut(parameter, combinations)
+	}
+	return combinations
+}
+
+// GetCombinationsCount returns the count of combinations of Parameters generated from the Matrix in PipelineTask.
+func (m *Matrix) GetCombinationsCount() int {
+	count := 0
+	if m.hasParams() {
+		count = 1
+		for _, param := range m.Params {
+			count *= len(param.Value.ArrayVal)
+		}
+	}
+	return count
+}
+
+func (m *Matrix) hasInclude() bool {
+	return m != nil && m.Include != nil && len(m.Include) > 0
+}
+
+func (m *Matrix) hasParams() bool {
+	return m != nil && m.Params != nil && len(m.Params) > 0
+}
+
+func fanOut(param Param, combinations []Params) []Params {
```

Comment:
Can you add docstrings for the functions in this file even though they're not exported?